src/utils/saml.ts

```diff
@@ -670,18 +672,104 @@ export const getPreValidateResponse = (
+      );
+      const AssertionNotBeforeDelta = pipe(
+        info.Assertion.getAttribute("NotBefore"),
+        NonEmptyString.decode,
+        E.chain(UTCISODateFromString.decode),
+        E.map(NotBefore => String(startTime - NotBefore.getTime())),
+        E.getOrElseW(() => InfoNotAvailable)
+      );
+      const AssertionSubjectNotOnOrAfterDelta = pipe(
+        info.Assertion.getElementsByTagNameNS(
+          SAML_NAMESPACE.ASSERTION,
+          "Subject"
+        ).item(0),
+        O.fromNullable,
+        O.chainNullableK(Subject =>
+          Subject.getElementsByTagNameNS(
+            SAML_NAMESPACE.ASSERTION,
+            "SubjectConfirmation"
+          ).item(0)
+        ),
+        O.chainNullableK(SubjectConfirmation =>
+          SubjectConfirmation.getElementsByTagNameNS(
+            SAML_NAMESPACE.ASSERTION,
+            "SubjectConfirmationData"
+          ).item(0)
+        ),
+        E.fromOption(() => new Error("Could not find elements")),
+        E.chainW(extractNotOnOrAfterDelta),
+        E.getOrElseW(() => InfoNotAvailable)
+      );
+      const AssertionConditionsNotOnOrAfterDelta = pipe(
+        info.Assertion.getElementsByTagNameNS(
+          SAML_NAMESPACE.ASSERTION,
+          "Conditions"
+        ).item(0),
+        O.fromNullable,
+        E.fromOption(() => new Error("Could not find elements")),
+        E.chainW(extractNotOnOrAfterDelta),
+        E.getOrElseW(() => InfoNotAvailable)
+      );
+
+      const timings = {
+        AssertionConditionsNotOnOrAfterDelta,
+        AssertionIssueInstantDelta,
+        AssertionNotBeforeDelta,
+        AssertionSubjectNotOnOrAfterDelta,
+        ResponseIssueInstantDelta
+      };
+
+      eventHandler({
+        data: {
+          idpIssuer,
+          message: "Deltas infos to determine a valid clockSkewMs",
+          requestId,
+          ...timings
+        },
+        name: "spid.info.deltas",
+        type: "INFO"
+      });
+    }
+  };
+
   pipe(
     responseElementValidationStep,
     TE.chain(returnRequestAndResponseStep),
     TE.chain(parseSAMLRequestStep),
     TE.chain(getIssueInstantFromRequestStep),
-    TE.chain(issueInstantValidationStep),
-    TE.chain(assertionIssueInstantValidationStep),
+    TE.chainFirst(issueInstantValidationStep),
+    TE.chainFirst(assertionIssueInstantValidationStep),
     TE.chain(authnContextClassRefValidationStep),
-    TE.chain(attributesValidationStep),
-    TE.chain(responseIssuerValidationStep),
-    TE.chain(assertionIssuerValidationStep),
-    TE.chainW(transformValidationStep),
+    TE.chainFirst(attributesValidationStep),
+    TE.chainFirst(responseIssuerValidationStep),
+    TE.chainFirst(assertionIssuerValidationStep),
+    TE.chainFirstW(transformValidationStep),
+    // log timings infos
+    TE.chainFirst(IssueInstantWithAuthnContextCR => {
+      extractAndLogTimings(IssueInstantWithAuthnContextCR);
+      return TE.right(void 0);
+    }),
```

Comment:
What if we change the `extractAndLogTimings` like :
```typescript
const extractAndLogTimings = (
    info: IIssueInstantWithAuthnContextCR
  ): TaskEither<Error | ITransformValidation, void> => {
```
```suggestion
    TE.chainFirst(extractAndLogTimings),
```